apollo-router/src/query_planner/caching_query_planner.rs

```diff
@@ -530,8 +561,11 @@ pub(crate) struct CachingQueryKey {
     pub(crate) hash: Arc<QueryHash>,
     pub(crate) metadata: CacheKeyMetadata,
     pub(crate) plan_options: PlanOptions,
+    pub(crate) config_mode: ConfigMode,
 }
 
+// Update this key every time the cache key or the query plan format has to change
```

Comment:
Yes fed version is included.

`CACHE_KEY_VERSION` seems like it might be a footgun that will hurt us if we rely on it, as it is pretty much guaranteed that we will forget to update it when we have to.

Federation version takes care of query plan format, so the sorts of things we have left are mutations to the query plan by plugins such as file uploads. This isn't an issue right now as query plan rewrite happens every time, but if we wanted to cache the rewrites then the possibility of us getting this right decreases rapidly. What if the config in file uploads changes the way that query plans are rewritten? This is a rabbit hole we don't want to go down. 

Once the rust query planner is in place and optimized it'd be good to revisit the entire query plan cache store to see how much benefit we get from caching, and if we still need it what sort of investment and limits we will place on it. For instance we could mandate that query plan rewrites will *never* be cached which would significantly reduce the scope of the problem.

Reply:
Maybe I'm misunderstand the purpose here. It could be just for debugging so that given a cache key we know how to interpret it.

Reply:
On the flip side, maybe we should have self describing keys: https://stackoverflow.com/questions/6320739/does-name-length-impact-performance-in-redis

Reply:
I added `CACHE_KEY_VERSION` following feedback on the other issue with schema hashing, to cover the case where we would change anything related to query plan caching and want to ignore the previously cached entries as a safety measure.